src/mscorlib/src/System/Runtime/InteropServices/NativeLibrary.cs

```diff
@@ -0,0 +1,347 @@
+        /// which has a direct dependency on the native library.</param>
+        /// <param name="paths">A flags value which contains the locations that should be searched during module load.</param>
+        /// <param name="result">A <see cref="NativeLibrary"/> instance that represents the loaded module,
+        /// or <see langword="null"/> if the module could not be found. Once loaded, a module cannot be unloaded.</param>
+        /// <returns><see langword="true"/> if the module was found and loaded, otherwise <see langword="false"/>.</returns>
+        /// <remarks>
+        /// There are some behavioral differences between this method and <see cref="DllImportAttribute"/>. For example, the
+        /// value provided in <paramref name="name"/> cannot be the display name of a managed assembly. Additionally, if
+        /// <paramref name="caller"/> is not specified, then <paramref name="paths"/> cannot have the flags values
+        /// <see cref="DllImportSearchPath.LegacyBehavior"/> or <see cref="DllImportSearchPath.AssemblyDirectory"/>.
+        /// If <paramref name="caller"/> is specified and <paramref name="paths"/> is <see cref="DllImportSearchPath.LegacyBehavior"/>
+        /// and an assembly-level <see cref="DefaultDllImportSearchPathsAttribute"/> is present, the assembly-level default
+        /// <see cref="DllImportSearchPath"/> value will be used instead.
+        /// This method may throw an exception in certain cases, such as if the module was located but is invalid
+        /// for the current architecture.
+        /// </remarks>
+        public static bool TryLoad(string name, Assembly caller, DllImportSearchPath paths, out NativeLibrary result)
+        {
+            // Parameter validation
+
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            RuntimeAssembly callerAsRuntimeAssembly = null;
+            if (caller != null)
+            {
+                callerAsRuntimeAssembly = caller as RuntimeAssembly;
+
+                if (callerAsRuntimeAssembly != null)
+                {
+                    throw new ArgumentException(
+                        paramName: nameof(caller),
+                        message: SR.Argument_MustBeRuntimeAssembly);
+                }
+
+                // Assembly-level default search paths win out over legacy behavior.
+                // This is a slight change in behavior from [DllImport], where search paths specified at
+                // the method level override search paths specified at the assembly level. But since our
+                // paths parameter isn't nullable, we need to resolve the ambiguity *somehow*, and this
+                // seems like the way that'll be least surprising to consumers.
+
+                if (paths == DllImportSearchPath.LegacyBehavior)
+                {
+                    var attr = caller.GetCustomAttribute<DefaultDllImportSearchPathsAttribute>();
```

Comment:
The unmanaged runtime has cache for this value to avoid reflection call every time.

Reply:
I'm going for simplicity here rather than performance. The action of having the OS load a module into the current process is by far going to dominate the run time of this method.